src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs

```diff
@@ -217,6 +235,43 @@ private async Task FillRequestLists(List<Hl7.Fhir.Model.Bundle.EntryComponent> b
             }
         }
 
+        public async Task ResolveBundleReferencesAsync(EntryComponent entry, Dictionary<string, string> referenceIdDictionary)
+        {
+            IEnumerable<ResourceReference> references = entry.Resource.GetAllChildren<ResourceReference>();
+
+            foreach (ResourceReference reference in references)
+            {
+                if (referenceIdDictionary.TryGetValue(reference.Reference, out var referenceId))
+                {
+                    reference.Reference = referenceId;
+                }
+                else
+                {
+                    if (reference.Reference.Contains("?", StringComparison.InvariantCulture))
+                    {
+                        string[] queries = reference.Reference.Split("?");
+                        string resourceType = queries[0];
```

Comment:
Not sure if we need to guard against a system wide search if it's possible here.

Reply:
Can you please elaborate this one?